196 basic outline (#199)
limebit/medmodels: docs/user_guide/02_medrecord_intro.md

# MedRecord

## Preface

> Just like every major library has its core class — [PyTorch](https://pytorch.org/) has `torch.Tensor`, [Numpy](https://numpy.org/) has `np.array` — MedModels revolves around `mm.MedRecord` as its central object.

Reviewer: I am not quite convinced with the wording, I would suggest:
```suggestion
Every major library has a central object that consitutes its core. For [PyTorch](https://pytorch.org/), it is the `torch.Tensor`, whereas for [Numpy](https://numpy.org/), it is the `np.array`. In our case, MedModels centres around the `mm.MedRecord` as its foundational structure.
```

MedModels delivers advanced data analytics methods out-of-the-box by utilizing a structured approach to data storage. This is enabled by the MedRecord class, which organizes data of any complexity within a graph structure. With its Rust backend implementation, MedRecord guarantees high performance, even when working with extremely large datasets.

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 3
---
```

## Adding Nodes to a MedRecord

Let's begin by introducing some sample medical data:

:::{list-table} Patients
:widths: 15 15 15 15
:header-rows: 1

*   - ID
    - Age
    - Sex
    - Loc
*   - Patient 01
    - 72
    - M
    - USA
*   - Patient 02
    - 74
    - M
    - USA
*   - Patient 03
    - 64
    - F
    - GER
:::

> This data, stored e.g. in Pandas looks like this:

Reviewer 2: feels like this is missing a comma or the e.g. should be in parentheses

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 5-13
---
```

> To work with this data in MedModels, we'll have to create a MedRecord. The MedRecord employes a builder pattern to easily construct complex datastructures. 

Reviewer 3: ```suggestion
To work with this data in MedModels, we'll have to create a MedRecord. The MedRecord employes a builder pattern to easily construct complex data structures. 
```

> The Builder Pattern simplifies creating complex objects by constructing them step by step. It improves flexibility, readability, and consistency, making it easier to manage and configure objects in a controlled way.

Reviewer 2: this does not explain how the code below works.

IMO it should say something like:

> In the example below we create a new medrecord using the builder pattern. We create a `MedRecordBuilder` and tell it to add the pandas dataframe as nodes with the column "ID" being the column we'll use for indexing. We also tell it to add these nodes into the group "Patients". ...

Reviewer 2: we should also explain somewhere (not in the medrecord module) that the builder pattern is an essential part of the library and can be used to create all major classes. (Not in this PR, just a thought)

Reviewer 2: @philippgrosser I agree but what I wanted to convey is that I think we need a general section somewhere (maybe just the user guide index page) that is about the similarities between all classes (every major class has a builder pattern that works in the same way, every data in-outuput is a medrecord, etc.)

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 43
---
```

> The MedModels MedRecord object, `record`, now contains three patients. Each patient is identified by an unique index and has specific attributes, such as age, sex, and location. These patients serve as the initial nodes in the graph structure of our MedRecord, and are represented as follows:

Reviewer 2: ```suggestion
The MedModels MedRecord object, `record`, now contains three patients. Each patient is identified by a unique index and has specific attributes, such as age, sex, and location. These patients serve as the initial nodes in the graph structure of our MedRecord, and are represented as follows:
```

```{image} https://raw.githubusercontent.com/limebit/medmodels-static/main/imgs/user_guide/02/02_medrecord_intro_01.png
:class: transparent-image
```
We can now proceed by adding additional data, such as the following medications.
```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 15-18
---
```

> Using the builder pattern to construct the MedRecord allows us to pass as many nodes and edges as needed. If nodes are not added during the initial graph construction, they can easily be added later to an existing MedRecord by calling `.add_nodes()`, where you provide the DataFrame and specify the column containing the node indices.

Reviewer 2: This I like because you explain the attributes for `add_nodes`

> Using the builder pattern to construct the MedRecord allows us to pass as many nodes and edges as needed. If nodes are not added during the initial graph construction, they can easily be added later to an existing MedRecord by calling `.add_nodes()`, where you provide the DataFrame and specify the column containing the node indices.

Reviewer 2: ```suggestion
Using the builder pattern to construct the MedRecord allows us to pass as many nodes and edges as needed. If nodes are not added during the initial graph construction, they can easily be added later to an existing MedRecord by calling `add_nodes`, where you provide the DataFrame and specify the column containing the node indices.
```

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python

> lines: 46

Reviewer: This python line does not add a "Medications" group, as portrayed in the image. It is misleading. Either we establish how to add groups now or afterwards. Maybe not add any groups at all until referring to them in section 4.

Author: This will be fixed with an update that is to be discussed today: The add_nodes function will get an extra argument to pass a group directly, then I will change this code.

:class: transparent-image
```
## Adding Groups to a MedRecord

For certain analyses, we may want to define specific subcohorts within our MedRecord for easier access. We can do this by defining named groups withing our MedRecored.

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 58
---
```
This group will include all the defined nodes, allowing for easier access during complex analyses. Both nodes and edges can be added to a group, with no limitations on group size. Additionally, nodes and edges can belong to multiple groups without restriction.

```{image} https://raw.githubusercontent.com/limebit/medmodels-static/main/imgs/user_guide/02/02_medrecord_intro_04.png
:class: transparent-image
```

## Saving and Loading MedRecords

When building a MedRecord, you may want to save it to create a persistent version. This can be done by storing it as a RON (Rusty Object Notation) file. The MedRecord can then be reloaded, allowing you to create a new instance from the saved RON file.

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 80-81
---
```

## Printing Overview Tables

> The MedModels MedRecord class is designed to efficiently handle large datasets while maintaining a standardized data structure that supports complex analysis methods. As a result, the structure within the MedRecord can become intricate and difficult to manage. To address this, MedModels offers tools to help keep track of the graph-based data. One such tool is the `print_node_overview` method:

Reviewer: We are referring to the method as `print_node_overview` but then we call a different one.

Author: Function will be renamed

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python
lines: 60
---
```

It will print an overview over all grouped nodes in the MedRecord.

```
------------------------------------------------------
Nodes Group Count Attribute Info                    
------------------------------------------------------
Diagnoses   1     ICD       Values: I.21             
                  ICDv      min: 10                  
                            max: 10                  
                            mean: 10.00              
Medications 2     Name      Values: Insulin, Wararin 
Patients    3     Age       min: 64                  
                            max: 74                  
                            mean: 70.00              
                  Loc       Values: GER, USA         
                  Sex       Values: F, M             
US-Patients 2     Age       min: 72                  
                            max: 74                  
                            mean: 73.00              
                  Loc       Values: USA              
                  Sex       Values: M                
------------------------------------------------------
```

> As shown, we have two groups of nodes—Patients and Medications—created when adding the nodes. Additionally, there’s a group called 'US-Patients' that we created. For each group of nodes, we can view their attributes along with a brief statistical summary, such as the minimum, maximum, and mean for numeric variables.

Reviewer 2: good explanation

> As shown, we have two groups of nodes—Patients and Medications—created when adding the nodes. Additionally, there’s a group called 'US-Patients' that we created. For each group of nodes, we can view their attributes along with a brief statistical summary, such as the minimum, maximum, and mean for numeric variables.

Reviewer 3: ```suggestion
As shown, we have two groups of nodes — Patients and Medications — created when adding the nodes. Additionally, there’s a group called 'US-Patients' that we created. For each group of nodes, we can view their attributes along with a brief statistical summary, such as the minimum, maximum, and mean for numeric variables.
```

> We can do the same to get an overview over edges in our MedRecord by using the `print_edge_overview`:

Reviewer 2: ```suggestion
We can do the same to get an overview over edges in our MedRecord by using the `print_edge_overview` method:
```

> As shown, we have two groups of nodes - Patients and Medications — created when adding the nodes. Additionally, there’s a group called 'US-Patients' that we created. For each group of nodes, we can view their attributes along with a brief statistical summary, such as the minimum, maximum, and mean for numeric variables.

Reviewer 2: ```suggestion
As shown, we have two groups of nodes - Patients and Medications - created when adding the nodes. Additionally, there’s a group called 'US-Patients' that we created. For each group of nodes, we can view their attributes along with a brief statistical summary, such as the minimum, maximum, and mean for numeric variables.
```

> We can do the same to get an overview over edges in our MedRecord by using the `.print_edge_overview()`:

Reviewer 2: ```suggestion
We can do the same to get an overview over edges in our MedRecord by using the `print_edge_overview` function:
```

> We can do the same to get an overview over edges in our MedRecord by using the `print_edge_overview` method:

Reviewer: We are referring to the method as `print_edge_overview` but then we call a different one.

Author: We will rename this function, this is also to be discussed today in MM weekly

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python

> lines: 62

Reviewer: Big problem! we are printing the same table twice (both nodes)

Author: Fixed, thank you!

---
```

```
------------------------------------------------------
Nodes Group Count Attribute Info                    
------------------------------------------------------
Diagnoses   1     ICD       Values: I.21             
                  ICDv      min: 10                  
                            max: 10                  
                            mean: 10.00              
Medications 2     Name      Values: Insulin, Wararin 
Patients    3     Age       min: 64                  
                            max: 74                  
                            mean: 70.00              
                  Loc       Values: GER, USA         
                  Sex       Values: F, M             
US-Patients 2     Age       min: 72                  
                            max: 74                  
                            mean: 73.00              
                  Loc       Values: USA              
                  Sex       Values: M                
------------------------------------------------------
```

## Accessing Elements in a MedRecord

> Now that we have stored some structured data in our MedRecord, we might want to access certain elements of it. The main way to do this is by either selecting the data with their indicies or via groups that they are in.

Reviewer 3: ```suggestion
Now that we have stored some structured data in our MedRecord, we might want to access certain elements of it. The main way to do this is by either selecting the data with their indices or via groups that they are in.
```

```{literalinclude} scripts/02_medrecord_intro.py
---
language: python

> lines: 64-78

Reviewer: Maybe better executing each cell separately? Or does it get too messy? This way no comments are needed

Author: yes, also thought so. The querying is also to be discussed in today's MM weekly

---
```
## Advanced Querying on a MedRecord

The MedRecord can be queried in very advanced ways in order to finde very specific nodes based on time, relations, neighbors or other. These advanced querying methods are covered in the next session of the user guide.

## Full example Code

> The full code examples for this chapter can be foud here: 

Reviewer 3: ```suggestion
The full code examples for this chapter can be found here: 
```

> The MedRecord can be queried in very advanced ways in order to finde very specific nodes based on time, relations, neighbors or other. These advanced querying methods are covered in the next session of the user guide.

Reviewer: ```suggestion
The MedRecord can be queried in very advanced ways in order to find very specific nodes based on time, relations, neighbors or other. These advanced querying methods are covered in the next session of the user guide.
```
```suggestion
The MedRecord can be queried in very advanced ways in order to finde very specific nodes based on time, relations, neighbors or other. These advanced querying methods are covered in the next session of the user guide.
```
```suggestion
The MedRecord can be queried in very advanced ways in order to finde very specific nodes based on time, relations, neighbors or other. These advanced querying methods are covered in the next session of the user guide.
```
